File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/makePayPalPayment/PayPalPaymentHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.makePayPalPayment

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.makePayPalPayment.PayPalPaymentApi
import features.makePayPalPayment.models.PayPalPaymentModel

import features.makePayPalPayment.serializePayPalPaymentState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class PayPalPaymentHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/make_pay_pal_payment"

    override val eventChannel: String
        get() = "customer_sdk/events/make_pay_pal_payment"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return makePayPalPayment(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun makePayPalPayment(call: MethodCall): String? {
            val result = PayPalPaymentApi().makePayPalPayment(
                intent = call.argument<String>("intent") as String,
                provider = call.argument<String>("provider") as String,
                giftcardId = call.argument<Int>("giftcardId") as Int,
                paypalReturnUrl = call.argument<String>("paypalReturnUrl") as String,
                paypalCancelUrl = call.argument<String>("paypalCancelUrl") as String
            ).toJsonResult<PayPalPaymentModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        PayPalPaymentApi().payPalPaymentState.serializePayPalPaymentState().observe { result ->
            pushSerializedResultToEventSink<PayPalPaymentApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/makePayPalPayment/PayPalPaymentHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class PayPalPaymentHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/make_pay_pal_payment"

    var eventChannel: String = "customer_sdk/events/make_pay_pal_payment"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await PayPalPaymentHandler.makePayPalPayment(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func makePayPalPayment(call: FlutterMethodCall) async -> String? {
        

        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
                        guard let intent = myArgs["intent"] as? String else {return nil}
                guard let provider = myArgs["provider"] as? String else {return nil}
                guard let giftcardId = myArgs["giftcardId"] as? Int32 else {return nil}
                guard let paypalReturnUrl = myArgs["paypalReturnUrl"] as? String else {return nil}
                guard let paypalCancelUrl = myArgs["paypalCancelUrl"] as? String else {return nil}

         let result = try await PayPalPaymentApi().makePayPalPayment(
                intent: intent,
                provider: provider,
                giftcardId: giftcardId,
                paypalReturnUrl: paypalReturnUrl,
                paypalCancelUrl: paypalCancelUrl
            ).serializeNative()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        PayPalPaymentApi().payPalPaymentState
            .serializePayPalPaymentState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
